prediction-applier/src/main/java/org/alfresco/hxi_connector/prediction_applier/hx_insight/HxPredictionReceiver.java

```diff
@@ -0,0 +1,135 @@
+import java.util.Queue;
+
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.camel.Exchange;
+import org.apache.camel.Processor;
+import org.apache.camel.builder.RouteBuilder;
+import org.apache.camel.component.jackson.JacksonDataFormat;
+import org.apache.camel.component.jackson.ListJacksonDataFormat;
+import org.springframework.stereotype.Component;
+
+import org.alfresco.hxi_connector.common.model.prediction.Prediction;
+import org.alfresco.hxi_connector.prediction_applier.config.PredictionListenerConfig;
+
+@Slf4j
+@Component
+@RequiredArgsConstructor
+@SuppressWarnings({"PMD.UnusedPrivateMethod", "PMD.UnusedFormalParameter", "PMD.LinguisticNaming", "PMD.LongVariable"})
+public class HxPredictionReceiver extends RouteBuilder
+{
+    private static final String PREDICTION_PROCESSOR_TRIGGER_ROUTE_ID = "prediction-processor-trigger-route";
+    private static final String PREDICTION_PROCESSOR_ROUTE_ID = "prediction-processor-route";
+    private static final String PREDICTION_PROCESSOR = "direct:prediction-processor";
+    private static final String IS_PREDICTION_PROCESSING_PENDING_KEY = "is-prediction-processing-pending";
+    private static final String HAS_NEXT_PAGE_KEY = "has-next-page";
+    private static final String PREDICTIONS_BATCH_KEY = "predictions-batch";
+
+    private final PredictionListenerConfig config;
+
+    @Override
+    public void configure()
+    {
+        // @formatter:off
+        JacksonDataFormat predictionsBatchDataFormat = new ListJacksonDataFormat(Prediction.class);
+        JacksonDataFormat predictionDataFormat = new JacksonDataFormat(Prediction.class);
+
+        from(config.predictionProcessorTriggerEndpoint())
+                .routeId(PREDICTION_PROCESSOR_TRIGGER_ROUTE_ID)
+                .choice()
+                .when(this::isProcessingPending)
+                    .log(DEBUG, log, "Prediction processing is pending, no need to trigger it")
+                .otherwise()
+                    .log(DEBUG, log, "Triggering prediction processing")
+                    .to(PREDICTION_PROCESSOR);
+
+        from(PREDICTION_PROCESSOR)
+                .routeId(PREDICTION_PROCESSOR_ROUTE_ID)
+                .process(setIsProcessingPending(true))
+                .loopDoWhile(this::hasNextPage)
+                    .log(DEBUG, log, "Fetching predictions")
+                    .to(config.predictionsSourceEndpoint())
+                    .log(DEBUG, log, "Sending predictions to internal buffer: ${body}")
+                    .unmarshal(predictionsBatchDataFormat)
+                    .process(this::savePredictionsBatch)
+                    .loopDoWhile(this::predictionsBatchNotEmpty)
+                        .process(this::setPredictionToSend)
+                        .marshal(predictionDataFormat)
+                        .log(TRACE, log, "Sending prediction to internal buffer: ${body}")
+                        .to(config.internalPredictionsBufferEndpoint())
```

Comment:
This route is so complicated that it's hard to understand what it does. I think that some pictograms in Javadoc would be really helpful 😅

Reply:
I really don't like javadoc tbh - I've added some but I would discuss if it is actually helpful or not